src/Trees/Generate.hs

```diff
@@ -46,13 +46,14 @@ genSynTree
   allowArrowOperators
   maxConsecutiveNegations
   minUniqueBinOps = do
-    nodes <- choose (minNodes, maxNodes) `suchThat` if hasNegations then const True else odd
-    sample <- syntaxShape nodes maxDepth allowArrowOperators hasNegations
-      `suchThat` \synTree ->
-        checkAtLeastOccurring synTree &&
-        checkMinUniqueOps synTree &&
-        treeDepth synTree >= minDepth &&
-        (not hasNegations || (consecutiveNegations synTree <= maxConsecutiveNegations))
+    sample <-
+      (do nodes <- choose (minNodes, maxNodes) `suchThat` if hasNegations then const True else odd
+          syntaxShape nodes maxDepth allowArrowOperators hasNegations
+        ) `suchThat` \synTree ->
+            checkAtLeastOccurring synTree &&
+            checkMinUniqueOps synTree &&
+            treeDepth synTree >= minDepth &&
+            (not hasNegations || (consecutiveNegations synTree <= maxConsecutiveNegations))
```

Comment:
You could try one by one moving the checks except `treeDepth synTree >= minDepth` back into the `do`-block, on the `syntaxShape` line, and see whether generation is still terminating.

Thus the distribution of generated trees over the `nodes` number would probably be improved a bit. Because now there are four reasons (lines 53-56) for which an already chosen `nodes` value is discarded, even though maybe only one of them is really a problem. Having more checks inside the `do` block would mean that more effort is made to satisfy a chosen `nodes` value. That is, `nodes` values that are harder to satisfy would ultimately "survive" more often.